### PR TITLE
[ENH] `prophet` based piecewise linear trend forecaster

### PR DESCRIPTION
**Pull request by sbuse**
#### Reference Issues/PRs

Discussion https://github.com/sktime/sktime/discussions/5197

Depends on https://github.com/sktime/sktime/pull/5622 which should be merged first.

#### What does this implement/fix? Explain your changes.

This PR extracts the piece wise linear trend from the prophet algorithm. The algorithm is well suited to be used for trend forecasting in a pipeline and less useful alone. 

#### Does your contribution introduce a new dependency? If yes, which one?

no new dependency, prophet was already implemented. 

#### What should a reviewer concentrate their feedback on?

Since this is my first PR for sktime it would be good if the reviewer could check if i have done everything required and according to the conventions. 

#### Did you add any tests for the change?

No I haven't added any test. 

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, I've added myself and possibly others to the [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) file - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [x] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

**Comment by sbuse** [comment]
Hello core developers this is my first PR and I would be happy to get some feedback and improve the code if needed.

File: sktime/forecasting/fbprophet.py
Comment (fkiraly) [inline sktime/forecasting/fbprophet.py:237]: there must be a newline at end of each file, linting will complain...

**Review by fkiraly — CHANGES_REQUESTED** [review]
Looks good on superficial look!

One issue - which the tests will also highlight - is that you are not adhering to sktime code formatting (linting). These are various small and - logic-wise - inconsequential things, but it is enforced.

Instructions on this, including how to set this up on your computer in an automated way, here:
https://www.sktime.net/en/stable/developer_guide/coding_standards.html

**Comment by sbuse** [comment]
> Looks good on superficial look!
> 
> One issue - which the tests will also highlight - is that you are not adhering to sktime code formatting (linting). These are various small and - logic-wise - inconsequential things, but it is enforced.
> 
> Instructions on this, including how to set this up on your computer in an automated way, here: https://www.sktime.net/en/stable/developer_guide/coding_standards.html

Thanks @fkiraly for the comments. I ran `pre-commit run --all-files` which hopefully fixed the all linting complaints.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Looks great!

There is a subtle issue, related to how `sktime` does testing.

In testing, we pass multiple data scenarios to fit/predict, this is with varying index and index type.

In `get_test_params`, you have specified changepoints that make sense for some indices, but not for others, but all data scenarios are run with these parameters, which can lead to the "changepoints outside range" issue.

An acceptable solution would be, imo, to specify changepoints only by number or relatively, without direct reference to the index range, and set up separate tests, in a `test_pwl_trend` or similar file, with changepoints matching index range.

**Review by fkiraly — DISMISSED** [review]
Looks good to me.

I would recommend, let's add at least a second parameter set to `get_test_params` as well, covering different parameter settings.

Will not merge yet as this is still a draft, but looks done.

**Comment by fkiraly** [comment]
FYI @sbuse , @tpvasconcelos, @yarnabrina, this PR actually contains an example of a second descendant of `_ProphetAdapter`.

Post-merge https://github.com/sktime/sktime/pull/5597, code that previously would not have failed, fails now.
To address this, I have decoupled the `fit_kwargs` parameter better in https://github.com/sktime/sktime/pull/5622, and have merged it into this PR. #5622 should be merged first.

Reviews would be appreciated, from the perspective of the prophet adapter.

File: sktime/forecasting/trend/_pwl_trend_forecaster.py
Comment (fkiraly) [inline sktime/forecasting/trend/_pwl_trend_forecaster.py:54]: doctests are always run, even if prophet is not present, so we should skip it - see with the `Prophet` forecaster how this is done.

There is an idiomatic `doctest` way which I do not remember - and, currently, skipif, and move the code to a`pytest` test is currently the most common in `sktime`.

**Comment by sbuse** [comment]
> Looks good to me.
> 
> I would recommend, let's add at least a second parameter set to `get_test_params` as well, covering different parameter settings.
> 
> Will not merge yet as this is still a draft, but looks done.

Okay, I will add `test_pwl_trend.py` and include some tests there. It feels a bit redundant since this is just a wrapper but I figure it is good practice to fully test the code you want to commit.

**Comment by tpvasconcelos** [comment]
@sbuse @fkiraly quick question on this one. Referencing the original example in https://github.com/sktime/sktime/discussions/5197 , isn't this already possible with:

```python
pwl_trend_forecaster = Prophet(
    daily_seasonality=False,
    weekly_seasonality=False,
    yearly_seasonality=False
)
forecaster = TransformedTargetForecaster(
    [
        ("detrend", Detrender(forecaster=pwl_trend_forecaster)),
        ("forecast", ARIMA() )  
    ]
)
```

**Comment by fkiraly** [comment]
> isn't this already possible with:

Perhaps, but (a) I do not know the `prophet` forecaster well enough, and (b) even if, I think it's a useful shorthand or special case.

Irrespective of this, the current policy is that we add what individual contributors and users find useful, acknowledging that there are different tastes and use case needs. Requirements are proper docs and crediting only.

I also think it is quite useful from maturing software architecture and testing around the `prophet` adapter, as the necessity of #5622 shows, although that's not a criterion anywhere.

**Comment by sbuse** [comment]
> @sbuse @fkiraly quick question on this one. Referencing the original example in #5197 , isn't this already possible with:
> 
> ```python
> pwl_trend_forecaster = Prophet(
>     daily_seasonality=False,
>     weekly_seasonality=False,
>     yearly_seasonality=False
> )
> forecaster = TransformedTargetForecaster(
>     [
>         ("detrend", Detrender(forecaster=pwl_trend_forecaster)),
>         ("forecast", ARIMA() )  
>     ]
> )
> ```

@tpvasconcelos  you are right, it is possible to force prophet to only do trend fitting. Unfortunately i wasn't aware of this and i thought it would be useful to add it to the catalog of detenders.

**Comment by tpvasconcelos** [comment]
@sbuse @fkiraly Thanks a lot for the quick replies! 

After thinking about this one for a bit more, there **is** actually a difference between the two approaches. Prophet uses [MLE](https://en.wikipedia.org/wiki/Maximum_likelihood_estimation) to internally optimise whatever model is picked via the high-level hyperparameters. So the current approach in this pull request will return the trend component for the following MLE optimised additive model: 

```
Y(t) = T(t) + S_daily(T) + S_weekly(T) + S_yearly(T) + e
```

, while the example I shared above will only try to fit the trend component (i.e., MLE on `Y(t) = T(t) + e`).

Taking this into account, the current implementation hides the additive seasonal terms from the user (and does not allow the user to toggle these or configure them in any way) which is a modeling assumption that might now work well for many use cases and give unexpected results. 

One option would be to default to `[daily|weekly|yearly]_seasonality=False` in this new `PiecewiseLinearTrendForecaster` estimator. However, for this simple use case, I still think that the existing `Prophet` interface that sktime exposes is enough and more explicit.

Another potential issue with this implementation is that it invites users to contribute yet another implementation for the logistic growth trend that Prophet also provides.

One option could be to introduce a new parameter to the existing `Prophet` constructor that allows users to extract specific components from the underlying Prophet model. Something like `extract_component=trend|weekly|yearly|holidays|etc...`. The default value for this parameter would be `"yhat"`. And the implementation would be very simple since we just have to pass this parameter here (where `"yhat"` is now):

https://github.com/sktime/sktime/blob/c540a77b8f307c61635690491b6121e576c63427/sktime/forecasting/base/adapters/_fbprophet.py#L197

i.e.,
```diff
-  y_pred = out.loc[:, "yhat"] 
+  y_pred = out.loc[:, self. extract_component] 
```

File: sktime/forecasting/trend/_pwl_trend_forecaster.py
Comment (tpvasconcelos) [inline sktime/forecasting/trend/_pwl_trend_forecaster.py:25]: This implementation will always automatically add two additive seasonal components (`sp=7` and `sp=365`) to the underlying model, even for if the underlying data is non-daily (since we're now passing the daily assumption to Prophet). And since the user can only pass trend-related hyperparameters to the constructor, there is no way to disable this behaviour. I believe this is a very unintuitive default behaviour to have in a class that appears to be a simple generic piecewise linear trend forecaster. As I mentioned in [this comment](https://github.com/sktime/sktime/pull/5592#issuecomment-1857745836), I belive that disabling all seasonal components in the underlying Prophet model (`[daily|weekly|yearly]_seasonality=False`) would solve this.

Comment (fkiraly) [reply sktime/forecasting/trend/_pwl_trend_forecaster.py:25]: so, let's do that? I did not expect seasonal components.

Comment (fkiraly) [reply sktime/forecasting/trend/_pwl_trend_forecaster.py:32]: I renamed the estimator and updated the docstring to clarifiy what happens with seasonality

**Comment by sbuse** [comment]
Hi @tpvasconcelos, Hi @fkiraly 
I ran an experimental test to see if the suggestion from @tpvasconcelos or the current implementation works better on an experimental level. The suggestion looks good but a quick test showed that the current implementation and the suggestions tend to not return the same result. That is why I wanted to test this more rigorously. I created some random data with no structure and let the two implementations fit it. What I got, as a result, is that the current implementation seems to be a tiny bit better than what @tpvasconcelos suggested. I think this shows that the current implementation is fine. Please let me know if you think otherwise or have any suggestions. 


``` python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sktime.forecasting.base                import ForecastingHorizon
from sktime.forecasting.trend               import ProphetPiecewiseLinearTrendForecaster
from sktime.forecasting.model_selection     import temporal_train_test_split
from sktime.forecasting.fbprophet           import Prophet
from sktime.performance_metrics.forecasting import MeanSquaredError

def Randomwalk1D(n): 
   y = 0
   yposition = [0] 
   for i in range(n-1):
       step = np.random.uniform(0,1)
       if step < 0.5:  
           y += 1  
       if step > 0.5:   
           y += -1 
       yposition.append(y)
   return yposition

n = 1000
np.random.seed(42)
mse = MeanSquaredError(square_root=True)
all_mse_current = []
all_mse_suggestion = []

for prior_scale in np.logspace(-3,-1/3,num=10): #values within 0.001 and 0.46
    for i in range(500):
        y = pd.Series(
                    data=Randomwalk1D(n),
                    index=pd.date_range('2016-01-01',periods=n, freq='15min')
                    )
        y_train, y_test = temporal_train_test_split(y)
        fh = ForecastingHorizon(y.index, is_relative=False)
        current = ProphetPiecewiseLinearTrendForecaster(
                            changepoint_prior_scale=prior_scale
                            )
        suggestion = Prophet(
                            daily_seasonality=False,
                            weekly_seasonality=False,
                            yearly_seasonality=False,
                            changepoint_prior_scale=prior_scale
                            )
        
        y_pred_current  = current.fit(y).predict(fh)
        y_pred_suggestion  = suggestion.fit(y).predict(fh)        
        mse_current = mse(y,y_pred_current)
        mse_suggestion = mse(y,y_pred_suggestion)

        all_mse_current.append(mse_current)
        all_mse_suggestion.append(mse_suggestion)

total = len(np.array(all_mse_current))
current_wins = np.sum(np.array(all_mse_current)<np.array(all_mse_suggestion))
suggestion_wins = np.sum(np.array(all_mse_current)>np.array(all_mse_suggestion))
ties = np.sum(np.array(all_mse_current)==np.array(all_mse_suggestion))

print(f"Out of {total} runs: \nThe current implementation is more accurate in "+
      f"{current_wins} ({np.round_(current_wins/total*100,1)}%) cases.\n"+
      f"The suggested change is more accurate in {suggestion_wins}"+
      f"({np.round_(suggestion_wins/total*100,1)}%) cases.\n"+
      f"In {ties} ({np.round_(ties/total*100,1)}%) cases they return the same value.")        
```
```print
Out of 5000 runs: 
The current implementation is more accurate in 2908 (58.2%) cases.
The suggested change is more accurate in 2092(41.8%) cases.
In 0 (0.0%) cases they return the same value.
        
```

**Comment by fkiraly** [comment]
probably will depend on the data - my intuition would tell me, if there are different sensible suggestions on how it should behave internally, just expose them to the user, i.e., the seasonality parameters, with default being current.

**Pull request merged**
